### PR TITLE
Unit test and minor fix

### PR DESCRIPTION
**Pull request by fiona-gladwin**
- Add support to test FileSourceReader with Brightness and CMN
- Add Brightness fixed support
- Grayscale fix for supported datalaoder

### CODE REVIEW

File: utilities/rocAL/rocAL_unittests/testAllScripts.sh [hunk 1]
```diff
@@ -26,7 +26,7 @@ caffe2_classification_path=${ROCAL_DATA_PATH}/rocal_data/caffe2/classification/
 caffe2_detection_path=${ROCAL_DATA_PATH}/rocal_data/caffe2/detection/
 mxnet_path=${ROCAL_DATA_PATH}/rocal_data/mxnet/
 output_path=../rocal_unittest_output_folder_$(date +%Y-%m-%d_%H-%M-%S)/
-golden_output_path=${ROCAL_DATA_PATH}/rocal_data/GoldenOutputs/
+golden_output_path=${ROCAL_DATA_PATH}/rocal_data/complete_tensor_golden_output/
```
Comment (fiona-gladwin) [inline utilities/rocAL/rocAL_unittests/testAllScripts.sh:29]: @Akilesh2 why was this changed
It should be TensorGoldenOutputs here please check

**Review by sampath1117 — COMMENTED** [review]
@Akilesh2 
Please address review comments

File: rocAL/rocAL/source/api/rocal_api_augmentation.cpp [hunk 2]
```diff
@@ -642,35 +642,39 @@ rocalBrightness(
...
     }
 
     auto context = static_cast<Context*>(p_context);
-    auto input = static_cast<Image*>(p_input);
-    try
-    {
-        output = context->master_graph->create_image(input->info(), is_output);
+    auto input = static_cast<Tensor*>(p_input);
+    try {
+        RocalTensorlayout op_tensorLayout = (RocalTensorlayout)rocal_tensor_output_layout;
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply rocAL/rocAL/source/api/rocal_api_augmentation.cpp]: Done

File: rocAL/rocAL/source/api/rocal_api_augmentation.cpp [hunk 3]
```diff
@@ -642,35 +642,39 @@ rocalBrightness(
...
+        RocalTensorDataType op_tensorDataType = (RocalTensorDataType)rocal_tensor_output_datatype;
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply rocAL/rocAL/source/api/rocal_api_augmentation.cpp]: Done changed to camel case

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp [hunk 4]
```diff
@@ -106,6 +106,40 @@ auto convert_decoder_mode= [](RocalDecodeDevice decode_mode)
...
+    switch(color_format) {
+        case ROCAL_COLOR_RGB24:
+        case ROCAL_COLOR_BGR24: {
+            std::vector<size_t> dims_vector = {n, h, w, c};
+            return std::make_tuple(RocalTensorlayout::NHWC, dims_vector);
+        }
+        case ROCAL_COLOR_U8:
+        case ROCAL_COLOR_RGB_PLANAR: {
+            std::vector<size_t> dims_vector = {n, c, h, w};
+            return std::make_tuple(RocalTensorlayout::NCHW, dims_vector);
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply rocAL/rocAL/source/api/rocal_api_data_loaders.cpp:119]: The color format is present so retaining the same

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp [hunk 5]
```diff
@@ -106,6 +106,40 @@ auto convert_decoder_mode= [](RocalDecodeDevice decode_mode)
     }
 };
 
+auto convert_color_format_to_layout = [](RocalImageColor color_format, size_t n, size_t h, size_t w, size_t c) {
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply rocAL/rocAL/source/api/rocal_api_data_loaders.cpp:109]: Not required. As Image has color. Tensor is generic

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 6]
```diff
@@ -189,7 +189,7 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
 
     // Creating uniformly distributed random objects to override some of the default augmentation parameters
     RocalIntParam color_temp_adj = rocalCreateIntParameter(-50);
-    RocalIntParam mirror = rocalCreateIntParameter(1);
+    RocalIntParam mirror = rocalCreateIntParameter(0);
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp:192]: I think the Golden outputs are generated with the mirror as 0

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 7]
```diff
@@ -199,12 +199,14 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
     bool no_crop = false;
 #endif
 
-    RocalImage input1;
+    RocalTensor input1;
+    RocalTensorLayout tensorLayout = (rgb != 0) ? RocalTensorLayout::ROCAL_NHWC : RocalTensorLayout::ROCAL_NCHW;
+    RocalTensorOutputType tensorOutputType = RocalTensorOutputType::ROCAL_UINT8;
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp]: The convention in unit test is rgb denotes multichannel or single channel input
And for single channel RPP follows NCHW so we have added the check here

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 8]
```diff
@@ -560,17 +562,17 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
...
+    break;*/
     case 25:
     {
         std::cout << ">>>>>>> Running "
                   << "rocalCropMirrorNormalize" << std::endl;
-        std::vector<float> mean;
-        std::vector<float> std_dev;
-        image1 = rocalCropMirrorNormalize(handle, image0, 1, 224, 224, 0.2, 0.2, 1, mean, std_dev, true, mirror);
+        std::vector<float> mean{0, 0, 0};
+        std::vector<float> std_dev{1, 1, 1};
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp]: chnaged

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 9]
```diff
@@ -842,25 +846,49 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
...
                         }
-                    }
+                    }*/
                     std::cout << "\n";
                 }
             }
             break;
             case 2: //detection pipeline
             {
+                // int img_size = rocalGetImageNameLen(handle, image_name_length);
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp]: Removed

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 10]
```diff
@@ -901,33 +929,51 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
...
         {
-            if(DISPLAY)
-            cv::imshow("output",mat_output);
+            auto output_tensor = output_tensor_list->at(idx);
+            int h = output_tensor->shape().at(1) * output_tensor->dims().at(0);
+            int w = output_tensor->shape().at(0);
+            mat_input = cv::Mat(h, w, cv_color_format);
+            mat_output = cv::Mat(h, w, cv_color_format);
+            
+            unsigned char *out_buffer = (unsigned char *)malloc(output_tensor->data_size());
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp]: Done

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 11]
```diff
@@ -842,25 +846,49 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
...
-                    std::cerr << "\n Number of box:  " << bb_label_count[i];
-                int bb_labels[size];
-                rocalGetBoundingBoxLabel(handle, bb_labels);
-                float bb_coords[size * 4];
-                rocalGetBoundingBoxCords(handle, bb_coords);
+                RocalTensorList bbox_labels = rocalGetBoundingBoxLabel(handle);
+                RocalTensorList bbox_coords = rocalGetBoundingBoxCords(handle);
+                for(int i = 0; i < bbox_labels->size(); i++)
+                {
+                    int * labels_buffer = (int *)(bbox_labels->at(i)->buffer());
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp]: Done

[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp]: Done

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 12]
```diff
@@ -199,12 +199,14 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
     bool no_crop = false;
 #endif
 
-    RocalImage input1;
+    RocalTensor input1;
```
Comment (shobana-mcw) [inline utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp:202]: Do we need to name it input1?

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 13]
```diff
@@ -901,33 +909,50 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
...
         compression_params.push_back(IMWRITE_PNG_COMPRESSION);
         compression_params.push_back(9);
 
-        mat_input.copyTo(mat_output(cv::Rect(col_counter * w, 0, w, h)));
-        std::string out_filename = std::string(outName) + ".png";   // in case the user specifies non png filename
-        if (display_all)
-            out_filename = std::string(outName) + std::to_string(index) + ".png";   // in case the user specifies non png filename
-
-        if (color_format == RocalImageColor::ROCAL_COLOR_RGB24)
+        for(int idx = 0; idx < output_tensor_list->size(); idx++)
```
Comment (shobana-mcw) [inline utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp:918]: Please add comments here.

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp [hunk 14]
```diff
@@ -2098,13 +2126,13 @@ rocalVideoFileSourceSingleShard(
...
-        std::vector<size_t> dims = {context->user_batch_size(), sequence_length, video_prop.height, 
-                                    video_prop.width, static_cast<unsigned>(num_of_planes)};
+        auto [tensor_format, dims] = convert_color_format_to_sequence_layout(rocal_color_format, context->user_batch_size(),
+                                                                             video_prop.height, video_prop.width, num_of_planes, sequence_length);
         auto info  = TensorInfo(std::move(dims),
                                      context->master_graph->mem_type(),
                                      RocalTensorDataType::UINT8);
         info.set_color_format(color_format);
-        info.set_tensor_layout(RocalTensorlayout::NFHWC);
+        info.set_tensor_layout(tensor_format);
```
Comment (sampath1117) [inline rocAL/rocAL/source/api/rocal_api_data_loaders.cpp:2135]: Can we rename the variable tensor_format to layout_format/tensor_layout?

File: utilities/rocAL/rocAL_unittests/testAllScripts.sh [hunk 15]
```diff
@@ -26,7 +26,7 @@ caffe2_classification_path=${ROCAL_DATA_PATH}/rocal_data/caffe2/classification/
 caffe2_detection_path=${ROCAL_DATA_PATH}/rocal_data/caffe2/detection/
 mxnet_path=${ROCAL_DATA_PATH}/rocal_data/mxnet/
 output_path=../rocal_unittest_output_folder_$(date +%Y-%m-%d_%H-%M-%S)/
-golden_output_path=${ROCAL_DATA_PATH}/rocal_data/GoldenOutputs/
+golden_output_path=${ROCAL_DATA_PATH}/rocal_data/GoldenOutputsTensor/
```
Comment (sampath1117) [inline utilities/rocAL/rocAL_unittests/testAllScripts.sh:29]: Do we have 2 golden output folders for batchpd and tensor?

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 16]
```diff
@@ -359,17 +361,17 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
 
     int resize_w = width, resize_h = height; // height and width
 
-    RocalImage image0 = input1;
-    // RocalImage image0 = rocalResize(handle, input1, resize_w, resize_h, false); // uncomment when processing images of different size
-    RocalImage image1;
+    RocalTensor image0 = input;
+    // RocalTensor image0 = rocalResize(handle, input, resize_w, resize_h, false); // uncomment when processing images of different size
```
Comment (sampath1117) [inline utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp:365]: Since we removed resize
Does the display of outputs works as expected? I remember this resize is done to avoid the overlapping of outputs of different sizes
Please check if applicable

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 17]
```diff
@@ -405,15 +407,15 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
                   << "rocalRotate" << std::endl;
         image1 = rocalRotate(handle, image0, true);
     }
-    break;
+    break;*/
```
Comment (sampath1117) [inline utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp:410]: Instead of add multiple /*, is it better use mutli line //?

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 18]
```diff
@@ -758,16 +762,16 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
...
                   << "rocalSSDRandomCrop" << std::endl;
-        image1 = rocalSSDRandomCrop(handle, input1, true);
+        image1 = rocalSSDRandomCrop(handle, input, true);
     }
-    break;
+    break;*/
     case 55:
     {
         std::cout << ">>>>>>> Running "
                   << "rocalCropMirrorNormalizeFixed_center crop" << std::endl;
```
Comment (sampath1117) [inline utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp:771]: Please check if we need to change the naming here

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 19]
```diff
@@ -802,34 +805,34 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
...
-        int label_id[inputBatchSize];
         int numOfClasses = 0;
         int image_name_length[inputBatchSize];
         switch(pipeline_type)
         {
             case 1: //classification pipeline
             {
-                rocalGetImageLabels(handle, label_id);
+                RocalTensorList labels = rocalGetImageLabels(handle);
+                int * label_id = reinterpret_cast<int *>(labels->at(0)->buffer());  // The labels are present contiguously in memory
```
Comment (sampath1117) [inline utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp:821]: Please remove the additional space after *

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 20]
```diff
@@ -802,34 +805,34 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
...
                 numOfClasses = num_of_classes;
                 int label_one_hot_encoded[inputBatchSize * numOfClasses];
                 rocalGetImageName(handle, img_name);
-                if (num_of_classes != 0)
+                /*if (num_of_classes != 0)
                 {
                     rocalGetOneHotImageLabels(handle, label_one_hot_encoded, numOfClasses,0);
-                }
+                }*/
                 std::cerr << "\nPrinting image names of batch: " << img_name<<"\n";
```
Comment (sampath1117) [inline utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp:831]: Add space before and after <<

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 21]
```diff
@@ -853,14 +856,18 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
...
-                rocalGetBoundingBoxCords(handle, bb_coords);
+                RocalTensorList bbox_labels = rocalGetBoundingBoxLabel(handle);
+                RocalTensorList bbox_coords = rocalGetBoundingBoxCords(handle);
+                for(int i = 0; i < bbox_labels->size(); i++) {
+                    int *labels_buffer = reinterpret_cast<int *>(bbox_labels->at(i)->buffer());
+                    float *bbox_buffer = reinterpret_cast<float *>(bbox_coords->at(i)->buffer());
+                    std::cerr << "\n>>>>> BBOX LABELS : ";
+                    for(int j = 0; j < bbox_labels->at(i)->dims().at(0); j++)
+                        std::cerr << labels_buffer[j] << " ";
+                    std::cerr << "\n>>>>> BBOXX : " <<bbox_coords->at(i)->dims().at(0) << " : \n";
```
Comment (sampath1117) [inline utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp:867]: Correct the spelling of BBOX

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 22]
```diff
@@ -901,33 +908,46 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
...
+            auto output_tensor = output_tensor_list->at(idx);
+            int h = output_tensor->shape().at(1) * output_tensor->dims().at(0);
+            int w = output_tensor->shape().at(0);
+            if(first_run) { // Allocate cv matrix and output buffers for each output in first run and reuse in every iteration
+                mat_input.emplace_back(cv::Mat(h, w, cv_color_format));
+                mat_output.emplace_back(cv::Mat(h, w, cv_color_format));
+            }
+            
+            output_tensor->copy_data(mat_input[idx].data, ROCAL_MEMCPY_HOST);
+            mat_input[idx].copyTo(mat_output[idx](cv::Rect(0, 0, w, h)));
```
Comment (sampath1117) [inline utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp:927]: There are 2 copies involved here, what does both of them do?
Add comments for better readability

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 23]
```diff
@@ -901,33 +908,46 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
...
+                if(DISPLAY)
+                    cv::imshow("output", mat_output[idx]);
+                else
+                    cv::imwrite(out_filename, mat_color, compression_params);
+            }
             else
-            cv::imwrite(out_filename, mat_output, compression_params);
+            {
+                if(DISPLAY)
+                cv::imshow("output", mat_output[idx]);
```
Comment (sampath1117) [inline utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp:944]: Aiign code under if and else condition

**Review by sampath1117 — COMMENTED** [review]
@fiona-gladwin 
Please address review comments

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp (hunk 12 above)
Comment (fiona-gladwin) [reply utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp]: Renamed to input

File: utilities/rocAL/rocAL_unittests/testAllScripts.sh (hunk 15 above)
Comment (fiona-gladwin) [reply utilities/rocAL/rocAL_unittests/testAllScripts.sh:29]: Yes

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp (hunk 16 above)
Comment (fiona-gladwin) [reply utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp:365]: Each output will be dumped as separate image so resize will not be necessary

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp (hunk 17 above)
Comment (fiona-gladwin) [reply utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp:410]: We need to add // to every line which will show more line changes
/* */ is better to use when commenting multiple lines

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 24]
```diff
@@ -901,33 +929,51 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
...
+        cv::Mat mat_input;
+        cv::Mat mat_output;
+        for(int idx = 0; idx < output_tensor_list->size(); idx++)
         {
-            if(DISPLAY)
-            cv::imshow("output",mat_output);
+            auto output_tensor = output_tensor_list->at(idx);
+            int h = output_tensor->shape().at(1) * output_tensor->dims().at(0);
+            int w = output_tensor->shape().at(0);
+            mat_input = cv::Mat(h, w, cv_color_format);
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp]: Added changes

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp (hunk 23 above)
Comment (fiona-gladwin) [reply utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp]: Done

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp (hunk 21 above)
Comment (fiona-gladwin) [reply utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp]: Done

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp (hunk 19 above)
Comment (fiona-gladwin) [reply utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp]: Done

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp (hunk 14 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/api/rocal_api_data_loaders.cpp]: Done
